### PR TITLE
Fit interior range to content

### LINKED ISSUE #713: Update inside content range to only contain actual content and not whitespace
* content range should shrink down to fit actual content
* the full interior should be removal range
* on empty interior content range should be the same as removal range

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Fixes #713 

## Checklist

- [ ] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)

### CODE REVIEW

**Comment by pokey** [comment]
Performed git surgery.  Original branch at https://github.com/cursorless-dev/cursorless/tree/bak/fitInteriorToContent; feel free to delete if this branch looks ok

File: src/processTargets/targets/InteriorTarget.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+export default class InteriorTarget extends BaseTarget {
+  insertionDelimiter = "";
+
+  constructor(parameters: CommonTargetParameters) {
+    super(parameters);
+  }
+
+  get contentRange() {
+    return shrinkRangeToFitContent(this.editor, this.state.contentRange);
```
Comment (pokey) [inline src/processTargets/targets/InteriorTarget.ts:12]: huh.  It's a bit confusing that the `contentRange` is actually the `removalRange` for this target 🤔.  Not sure I have a better solution, but it's a bit surprising.  Maybe `InteriorTarget` is the exception to `CommonTargetParameters` containing `contentRange`?

Let's Discord this one when you're back

**Review by pokey — CHANGES_REQUESTED** [review]
Nice direction; left some comments

File: src/util/selectionUtils.ts [hunk 2]
```diff
@@ -42,3 +42,58 @@ export function selectionFromRange(isReversed: boolean, range: Range) {
   const { start, end } = range;
   return isReversed ? new Selection(end, start) : new Selection(start, end);
 }
+
+export function shrinkRangeToFitContent(editor: TextEditor, range: Range) {
+  const { lineAt } = editor.document;
+
+  const start = (() => {
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/util/selectionUtils.ts:49]: That's definitely one way of doing it. I think both will give the same results but the current solution is probably more performance effective

[the remark this answers is not in the document]
Comment (pokey) [reply src/util/selectionUtils.ts:49]: Maybe?  We don't know how `lineAt` and `line.text` work.  I would be inclined to avoid premature optimisation if we can shed 40 lines of code

File: src/processTargets/targets/InteriorTarget.ts (hunk 1 above)
Comment (pokey) [reply src/processTargets/targets/InteriorTarget.ts:12]: I'd be tempted to have something like `definingRange` as the common parameter, and then `contentRange` by default returns it, but things can override it

Comment (AndreasArvidsson) [reply src/processTargets/targets/InteriorTarget.ts:12]: We can do that

File: src/util/selectionUtils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/util/selectionUtils.ts]: Updated. Is the current solution better?

[the remark this answers is not in the document]
Comment (pokey) [reply src/util/selectionUtils.ts]: Much cleaner!

File: src/util/selectionUtils.ts [hunk 3]
```diff
@@ -42,3 +42,15 @@ export function selectionFromRange(isReversed: boolean, range: Range) {
   const { start, end } = range;
   return isReversed ? new Selection(end, start) : new Selection(start, end);
 }
+
+export function shrinkRangeToFitContent(editor: TextEditor, range: Range) {
```
Comment (pokey) [inline src/util/selectionUtils.ts:46]: Docstring please 😊

File: src/processTargets/targets/InteriorTarget.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/processTargets/targets/InteriorTarget.ts:12]: You want me to go ahead and do that change in this pr?

Comment (pokey) [reply src/processTargets/targets/InteriorTarget.ts:12]: Might be worth a quick convo

Comment (AndreasArvidsson) [reply src/processTargets/targets/InteriorTarget.ts:12]: Sounds good

**Review by pokey — COMMENTED** [review]
- [ ] Stop using `CommonTargetParameters` for parameters to `InteriorTarget`; instead use `fullInteriorRange` and then construct `contentRange` before passing to `BaseTarget`

File: src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml [hunk 4]
```diff
@@ -23,12 +23,13 @@ initialState:
   marks: {}
 finalState:
   documentContents: |-
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: do we have a test for "chuck head core curly"?  If not, worth adding I think

File: src/util/selectionUtils.ts [hunk 5]
```diff
@@ -42,3 +42,22 @@ export function selectionFromRange(isReversed: boolean, range: Range) {
   const { start, end } = range;
   return isReversed ? new Selection(end, start) : new Selection(start, end);
 }
+
+/**
+ * Shrink given range to fit text content.
```
Comment (pokey) [inline src/util/selectionUtils.ts:47]:
```suggestion
 * Return a copy of {@link range} excluding any leading or trailing whitespace.
```

File: src/util/selectionUtils.ts [hunk 6]
```diff
@@ -42,3 +42,22 @@ export function selectionFromRange(isReversed: boolean, range: Range) {
...
+ * If the given range has no text content the given range will be returned unchanged.
```
Comment (pokey) [inline src/util/selectionUtils.ts:48]: What does "no text content" mean?  Does that mean "if the range contains only whitespace"?

**Review by pokey — COMMENTED** [review]
Basically there!  In addition to a few minor comments inline:

- [ ] Would be good to add a "change" test where the core is empty
- [ ] Would be good to add a "change" test where the core includes only whitespace

File: src/util/selectionUtils.ts (hunk 6 above)
Comment (AndreasArvidsson) [reply src/util/selectionUtils.ts:48]: That or empty range

File: src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml (hunk 4 above)
Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: It will remove the same range since we are only looking at the content ranges in head/tail stage.

Comment (pokey) [reply src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: Hmm that is a bit unfortunate. I'm not entirely sure what is the right thing to do in that situation. I'm not sure there's anything else we can do reliably, but I was kind of expecting it to include the white space

Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: In that case we probably need to pass removal and content range to the head/tail stage. Or pass the leading and trailing delimiters or something like that.

Comment (pokey) [reply src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: Yeah I'm just not sure there's a reliable way to do it without all targets needing to know something about head and tail. Because targets need to be able to be smart about constructing removal range, and if head is just taking it and slicing it in half, that could violate the logic, eg we could possibly end up merging tokens

I don't think there's a simple way to do better than what you have here. Unless you have an idea?

Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: Unfortunately I don't have any good answers right now

Comment (pokey) [reply src/test/suite/fixtures/recorded/selectionTypes/clearHeadCoreCurly.yml:25]: No worries I'm happy to ship it and see if we find ourselves wanting this. Head tail does return a token target so at least tries to do some cleanup. Let's see if that's good enough
